File: database_di.py
# database_di.py (120 lines)
from abc import ABC, abstractmethod

class Database(ABC):
    @abstractmethod
    def execute_query(self, query, params=None):
        pass

class PostgreSQLDatabase(Database):
    def __init__(self, dbname, user, password, host):
        import psycopg2
        self.connection = psycopg2.connect(dbname=dbname, user=user, password=password, host=host)

    def execute_query(self, query, params=None):
        cursor = self.connection.cursor()
        cursor.execute(query, params or ())
        return cursor

# DI: Abstract class allows swapping databases (e.g., SQLite, MySQL).

class UserService:
    def __init__(self, db: Database):
        self.db = db

    def get_user(self, user_id):
        cursor = self.db.execute_query("SELECT * FROM users WHERE id = %s", (user_id,))
        return cursor.fetchone()

# Decoupled: DB implementation can change without modifying UserService.

File: database_di.js
// database_di.js (150 lines)
class Database {
    executeQuery(query, params) {
        throw new Error("Method not implemented");
    }
}

class MySQLDatabase extends Database {
    constructor(config) {
        super();
        const mysql = require('mysql');
        this.connection = mysql.createConnection(config);
    }

    executeQuery(query, params) {
        return new Promise((resolve, reject) => {
            this.connection.query(query, params, (err, results) => {
                if (err) reject(err);
                resolve(results);
            });
        });
    }
}

class UserService {
    constructor(database) {
        this.database = database;
    }

    async getUser(userId) {
        return this.database.executeQuery('SELECT * FROM users WHERE id = ?', [userId]);
    }
}

// DI: Database implementation can be swapped (e.g., SQLiteDatabase).
